src/QueryPipeline/RemoteQueryExecutor.cpp

```diff
@@ -249,7 +254,19 @@ void RemoteQueryExecutor::sendQuery(ClientInfo::QueryKind query_kind)
             connections->sendIgnoredPartUUIDs(duplicated_part_uuids);
     }
 
-    connections->sendQuery(timeouts, query, query_id, stage, modified_client_info, true);
+    auto user = context->getAccessControl().read<User>(modified_client_info.initial_user, true);
+    boost::container::flat_set<String> granted_roles;
+    if (user)
+    {
+        const auto & access_control = context->getAccessControl();
+        for (const auto & e : user->granted_roles.getElements())
+        {
+            auto names = access_control.readNames(e.ids);
+            granted_roles.insert(names.begin(), names.end());
+        }
+    }
```

Comment:
This part looks strange? What is it for?

Reply:
Roles UUIDs might be different on cluster nodes, while names 100% match, so it makes sense to send roles as list of names rather than list of UUIDs.